### ISSUE TITLE
[message] could cause OOS if option visibility differs per-client

### ISSUE THREAD

**Issue by CelticMinstrel** [wesnoth/wesnoth#3477]
Thanks to @sevu for thinking of this.

The `[message]` code first filters the options list for those options that should be displayed (depending on the `[show_if]` conditions). It then synchronizes your choice based on the index of the chosen option in the filtered list.

If one of the `[show_if]` conditions depended on something that is not the same on all clients, such as a persistent global variable, the filtered list would be different on each client which could lead to an out-of-sync error.

To fix this I think we'd need some sort of two-way mapping between the filtered and unfiltered option indices - after the choice is returned, map it to the unfiltered index, synchronize that, then map it back after the choice is synchronized.

**Comment by Pentarctagon** [comment]
What about setting the indexes for the options prior to filtering them, and then synchronizing the choice on the unfiltered list?

**Comment by gfgtdf** [comment]
as a workaround or someone that doesn't want to do lua (with using lua you can just ship your addon with a fixed vertsion of message) you could do the follwoing construct as a workaround (manual syncing):

```
[unsynced]
  [if]
    ## check whether current side is local
    [then]
    [message]
        [option]
          message="option1 ..."
          [command]
            {VARIABLE chosen_option 1}
          [/command]
        [/option]
        [option]
          message="option1 ..."
          [command]
            {VARIABLE chosen_option 1}
          [/command]
        [/option]
        [option]
          message="option2 ..."
          [command]
            {VARIABLE chosen_option 2}
          [/command]
        [/option]
        [option]
          message="option3 ..."
          [command]
            {VARIABLE chosen_option 4}
          [/command]
        [/option]
      [/message]
    [/then]
  [/if]
[/unsynced]
[sync_variable]
  name=chosen_option
[/sync_variable]
[switch]
  variable=chosen_option
  [case]
  [/case]
  [case]
  [/case]
  [case]
  [/case]
....
[switch]
```

**Comment by CelticMinstrel** [comment]
> (with using lua you can just ship your addon with a fixed vertsion of message)

Ahahahahaha yeah, that's not a great idea. Sure you _can_ do it, but I'm fairly sure that's gonna require duplicating the entire message tag logic in your addon. Not the best of ideas. But sure, if you _really_ need it I suppose you could do it.

**Comment by gfgtdf** [comment]
> but I'm fairly sure that's gonna require duplicating the entire message tag logic in your addon

yes this basicially means copying one lua file from core to your addon, not a big deal.
